Doc/library/sys.monitoring.rst

```diff
@@ -0,0 +1,282 @@
+
+Tool identifiers
+----------------
+
+A tool identifier is an integer and associated name.
+Tool identifiers are used to discourage tools from interfering with each
+other and to allow multiple tools to operate at the same time.
+Currently tools are completely independent and cannot be used to
+monitor each other. This restriction may be lifted in the future.
+
+Before registering or activating events, a tool should choose an identifier.
+Identifiers are integers in the range 0 to 5.
+
+Registering and using tools
+'''''''''''''''''''''''''''
+
+.. function:: use_tool_id(id: range(6), name: str) -> None
+
+   Must be called before ``id`` can be used.
+   Raises a ``ValueError`` if ``id`` is in use.
+
+.. function:: free_tool_id(id: range(6)) -> None
+
+   Should be called once a tool no longer requires ``id``.
+
+.. function:: get_tool(id: range(6)) ->  str | None
+
+   Returns the name of the tool if ``id`` is in use,
+   otherwise it returns ``None``.
+
+All IDs are treated the same by the VM with regard to events, but the
+following IDs are pre-defined to make co-operation of tools easier::
+
+  sys.monitoring.DEBUGGER_ID = 0
+  sys.monitoring.COVERAGE_ID = 1
+  sys.monitoring.PROFILER_ID = 2
+  sys.monitoring.OPTIMIZER_ID = 5
+
+There is no obligation to set an ID, nor is there anything preventing a tool
+from using an ID even it is already in use.
+However, tools are encouraged to use a unique ID and respect other tools.
+
+Events
+------
+
+The following events are supported:
+
+* BRANCH: A conditional branch is taken (or not).
+* CALL: A call in Python code (event occurs before the call).
+* C_RAISE: Exception raised from any callable, except Python functions (event occurs after the exit).
+* C_RETURN: Return from any callable, except Python functions (event occurs after the return).
+* EXCEPTION_HANDLED: An exception is handled.
+* INSTRUCTION: A VM instruction is about to be executed.
+* JUMP: An unconditional jump in the control flow graph is made.
+* LINE: An instruction is about to be executed that has a different line number from the preceding instruction.
+* PY_RESUME: Resumption of a Python function (for generator and coroutine functions), except for throw() calls.
+* PY_RETURN: Return from a Python function (occurs immediately before the return, the callee's frame will be on the stack).
+* PY_START: Start of a Python function (occurs immediately after the call, the callee's frame will be on the stack)
+* PY_THROW: A Python function is resumed by a throw() call.
+* PY_UNWIND:  Exit from a Python function during exception unwinding.
+* PY_YIELD: Yield from a Python function (occurs immediately before the yield, the callee's frame will be on the stack).
+* RAISE: An exception is raised, except those that cause a ``STOP_ITERATION`` event.
+* RERAISE: An exception is re-raised, for example at the end of a ``finally`` block.
+* STOP_ITERATION -- An artificial ``StopIteration`` is raised;
+  see `the STOP_ITERATION event`_.
```

Comment:
A definition list might be more semantically appropriate here (though if you'd prefer to keep the list, please could we change ``STOP_ITERATION -- `` to ``STOP_ITERATION: ``?)

```suggestion
BRANCH
  A conditional branch is taken (or not).
CALL
  A call in Python code (event occurs before the call).
C_RAISE
  Exception raised from any callable, except Python functions (event occurs after the exit).
C_RETURN
  Return from any callable, except Python functions (event occurs after the return).
EXCEPTION_HANDLED
  An exception is handled.
INSTRUCTION
  A VM instruction is about to be executed.
JUMP
  An unconditional jump in the control flow graph is made.
LINE
  An instruction is about to be executed that has a different line number from the preceding instruction.
PY_RESUME
  Resumption of a Python function (for generator and coroutine functions), except for throw() calls.
PY_RETURN
  Return from a Python function (occurs immediately before the return, the callee's frame will be on the stack).
PY_START
  Start of a Python function (occurs immediately after the call, the callee's frame will be on the stack)
PY_THROW
  A Python function is resumed by a throw() call.
PY_UNWIND
  Exit from a Python function during exception unwinding.
PY_YIELD
  Yield from a Python function (occurs immediately before the yield, the callee's frame will be on the stack).
RAISE
  An exception is raised, except those that cause a ``STOP_ITERATION`` event.
RERAISE
  An exception is re-raised, for example at the end of a ``finally`` block.
STOP_ITERATION
  An artificial ``StopIteration`` is raised; see `the STOP_ITERATION event`_.
```